ext/pcntl/pcntl.c

```diff
@@ -767,81 +767,135 @@ PHP_FUNCTION(pcntl_sigprocmask)
 
 #ifdef HAVE_STRUCT_SIGINFO_T
 # if defined(HAVE_SIGWAITINFO) && defined(HAVE_SIGTIMEDWAIT)
-static void pcntl_sigwaitinfo(INTERNAL_FUNCTION_PARAMETERS, int timedwait) /* {{{ */
-{
-	zval            *user_set, *user_signo, *user_siginfo = NULL;
-	zend_long             tv_sec = 0, tv_nsec = 0;
-	sigset_t         set;
-	int              signo;
-	siginfo_t        siginfo;
-	struct timespec  timeout;
-
-	if (timedwait) {
-		ZEND_PARSE_PARAMETERS_START(1, 4)
-			Z_PARAM_ARRAY(user_set)
-			Z_PARAM_OPTIONAL
-			Z_PARAM_ZVAL(user_siginfo)
-			Z_PARAM_LONG(tv_sec)
-			Z_PARAM_LONG(tv_nsec)
-		ZEND_PARSE_PARAMETERS_END();
-	} else {
-		ZEND_PARSE_PARAMETERS_START(1, 2)
-			Z_PARAM_ARRAY(user_set)
-			Z_PARAM_OPTIONAL
-			Z_PARAM_ZVAL(user_siginfo)
-		ZEND_PARSE_PARAMETERS_END();
+
+static bool php_pcntl_process_user_signal_infos(
+	/* const */ HashTable *const user_signals,
+	sigset_t *const set
+) {
+	if (zend_hash_num_elements(user_signals) == 0) {
+		zend_argument_value_error(1, "cannot be empty");
+		return false;
 	}
 
-	if (sigemptyset(&set) != 0) {
+	errno = 0;
+	if (sigemptyset(set) != 0) {
 		PCNTL_G(last_error) = errno;
 		php_error_docref(NULL, E_WARNING, "%s", strerror(errno));
-		RETURN_FALSE;
+		return false;
 	}
 
-	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(user_set), user_signo) {
-		signo = zval_get_long(user_signo);
-		if (sigaddset(&set, signo) != 0) {
+	zval *user_signal_no;
+	ZEND_HASH_FOREACH_VAL(user_signals, user_signal_no) {
+		bool failed = true;
+		zend_long tmp = zval_try_get_long(user_signal_no, &failed);
+
+		if (failed) {
+			zend_argument_type_error(1, "signals must be of type int, %s given", zend_zval_value_name(user_signal_no));
+			return false;
+		}
+		/* Signals are positive integers */
+		if (tmp < 0 || tmp > INT_MAX) {
```

Comment:
what about `tmp < 1 || tmp > NSIG-1` ?

Reply:
> what about `tmp < 1 || tmp > NSIG-1` ?

I don't really know signals, but that probably makes more sense yes.

Reply:
Actually seems the "correct way" for us to check would be to use ``PCNTL_G(num_signals)``

Reply:
> What's the difference between `NSIG-1` and `SIGRTMAX`? as both seems to indicate the last valid signal.

On Linux yes not on freebsd.

Reply:
> Actually seems the "correct way" for us to check would be to use `PCNTL_G(num_signals)`

Reading about it indeed it s bette.